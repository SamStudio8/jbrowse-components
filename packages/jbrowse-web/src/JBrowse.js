import React from 'react'
import ReactDOM from 'react-dom'
import * as mst from 'mobx-state-tree'
import { Provider } from 'mobx-react'

import App from './ui/App'
import RootModelFactory from './RootModelFactory'
import HierarchicalTrackSelectorDrawerWidgetPlugin from './plugins/HierarchicalTrackSelectorDrawerWidget'
import BamAdapterPlugin from './plugins/BamAdapter'
import AlignmentsTrackPlugin from './plugins/AlignmentsTrack'
import LinearGenomeViewPlugin from './plugins/LinearGenomeView'

import {
  DrawerWidgetType,
  RendererType,
  ViewType,
  TrackType,
  AdapterType,
} from './Plugin'

const corePlugins = [
  HierarchicalTrackSelectorDrawerWidgetPlugin,
  BamAdapterPlugin,
  LinearGenomeViewPlugin,
  AlignmentsTrackPlugin,
]

// little helper class that keeps groups of callbacks that are
// then run in a specified order by group
class PhasedScheduler {
  phaseCallbacks = {}

  constructor(...phaseOrder) {
    this.phaseOrder = phaseOrder
  }

  add(phase, callback) {
    if (!this.phaseOrder.includes(phase))
      throw new Error(`unknown phase ${phase}`)
    if (!this.phaseCallbacks[phase]) this.phaseCallbacks[phase] = []
    this.phaseCallbacks[phase].push(callback)
  }

  run() {
    this.phaseOrder.forEach(phaseName => {
      if (this.phaseCallbacks[phaseName])
        this.phaseCallbacks[phaseName].forEach(callback => callback())
    })
  }
}

// the main class used to configure and start a new JBrowse app
class JBrowse {
  plugins = []

  elementTypes = {}

  elementCreationSchedule = new PhasedScheduler(
    'renderer',
    'adapter',
    'track',
    'view',
    'drawer widget',
  )

  workerGroups = {}

  typeBaseClasses = {
    'drawer widget': DrawerWidgetType,
    renderer: RendererType,
    track: TrackType,
    view: ViewType,
    adapter: AdapterType,
  }

  static lib = { 'mobx-state-tree': mst, React }

  constructor() {
    this.lib = JBrowse.lib

    this.getViewType = this.getElementType.bind(this, 'view')
    this.getTrackType = this.getElementType.bind(this, 'track')
    this.getAdapterType = this.getElementType.bind(this, 'adapter')
    this.getDrawerWidgetType = this.getElementType.bind(this, 'drawer widget')
    this.getRendererType = this.getElementType.bind(this, 'renderer')
    this.addTrackType = this.addElementType.bind(this, 'track')
    this.addViewType = this.addElementType.bind(this, 'view')
    this.addAdapterType = this.addElementType.bind(this, 'adapter')
    this.addDrawerWidgetType = this.addElementType.bind(this, 'drawer widget')
    this.addRendererType = this.addElementType.bind(this, 'renderer')

    // add all the core plugins
    corePlugins.forEach(PluginClass => {
      this.addPlugin(new PluginClass())
    })
  }

  addPlugin(plugin) {
    if (this.started)
      throw new Error('JBrowse already configured, cannot add plugins')
    if (this.plugins.includes(plugin))
      throw new Error(`plugin already installed`)
    plugin.install(this)
    this.plugins.push(plugin)
    return this
  }

  addElementType(groupName, creationCallback) {
    if (typeof creationCallback !== 'function')
      throw new Error('must provide a callback function')
    const typeBaseClass = this.typeBaseClasses[groupName]
    if (!typeBaseClass)
      throw new Error(`unknown pluggable element type ${groupName}, cannot add`)
    if (!this.elementTypes[groupName]) this.elementTypes[groupName] = {}

    this.elementCreationSchedule.add(groupName, () => {
      const element = creationCallback(this)
      if (this.elementTypes[groupName][element.name])
        throw new Error(
          `${groupName} ${
            element.name
          } already registered, cannot register it again`,
        )

      this.elementTypes[groupName][element.name] = element
    })

    return this
  }

  getElementType(groupName, typeName) {
    if (!typeName) throw new Error(`invalid type name "${typeName}"`)
    return (this.elementTypes[groupName] || {})[typeName]
  }

  getElementTypesInGroup(groupName) {
    return Object.values(this.elementTypes[groupName] || {})
  }

  getElementTypeMembers(groupName, memberName) {
    return this.getElementTypesInGroup(groupName).map(t => t[memberName])
  }

  configure() {
    // run the creation callbacks for each element type in order.
    // currently tracks, then views
    this.elementCreationSchedule.run()
    delete this.elementCreationSchedule

    const RootModel = RootModelFactory(this)
    this.model = RootModel.create({})

    this.plugins.forEach(plugin => plugin.configure())

    this.configured = true

    // console.log(JSON.stringify(getSnapshot(model)))

    return this
  }

  addWorkers(groups) {
    Object.entries(groups).forEach(([groupName, workers]) => {
      if (!this.workerGroups[groupName]) this.workerGroups[groupName] = []
      this.workerGroups[groupName].push(...workers)
    })
  }

  getWorkerGroup(name) {
    return this.workerGroups[name]
  }

  start() {
    ReactDOM.render(
      <Provider rootModel={this.model}>
        <App
          getViewType={this.getViewType}
          getDrawerWidgetType={this.getDrawerWidgetType}
        />
      </Provider>,
      document.getElementById('root'),
    )
    return this
  }
}

export default JBrowse
