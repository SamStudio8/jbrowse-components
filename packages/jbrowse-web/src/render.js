import Rpc from '@librpc/web'

import SimpleFeature from './util/feature'

// import { isStateTreeNode, getSnapshot } from 'mobx-state-tree'

export function fog() {}

// // given an object of things that might be MST models, snapshots any of them
// // that are not already snapshots
// function ensureSnapshots(things) {
//   const data = {}
//   Object.entries(things).forEach(([key, value]) => {
//     if (isStateTreeNode(value)) {
//       data[key] = getSnapshot(value)
//     } else {
//       data[key] = value
//     }
//   })
//   return data
// }

const getClient = pluginManager =>
  new Rpc.Client({
    workers: pluginManager.getWorkerGroup('render'),
  })

export async function renderRegionWithWorker(pluginManager, args) {
  const result = await getClient(pluginManager).call('renderRegion', args, {
    timeout: args.timeout,
  })

  // convert the feature JSON to SimpleFeature objects
  result.features = result.featureJSON.map(j => SimpleFeature.fromJSON(j))
  return result
}

export async function freeSessionResourcesInWorker(pluginManager, sessionId) {
  return getClient(pluginManager).call('freeSessionResources', sessionId)
}
