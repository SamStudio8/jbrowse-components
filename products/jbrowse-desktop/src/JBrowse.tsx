import React, { useEffect, useState, Suspense } from 'react'
import { observer } from 'mobx-react'
import { CssBaseline, ThemeProvider } from '@material-ui/core'
import { getConf } from '@jbrowse/core/configuration'
import { App, createJBrowseTheme } from '@jbrowse/core/ui'
import PluginManager from '@jbrowse/core/PluginManager'
import { AssemblyManager } from '@jbrowse/plugin-data-management'
import deepmerge from 'deepmerge'

// locals
import { RootModel } from './rootModel'

const JBrowse = observer(
  ({ pluginManager, mode }: { mode: string; pluginManager: PluginManager }) => {
    const { rootModel } = pluginManager

    return rootModel ? (
      <JBrowseNonNullRoot mode={mode} rootModel={rootModel as RootModel} />
    ) : null
  },
)

const JBrowseNonNullRoot = observer(
  ({ rootModel, mode }: { mode: string; rootModel: RootModel }) => {
    const [firstLoad, setFirstLoad] = useState(true)
    const { session, jbrowse, error, isAssemblyEditing, setAssemblyEditing } =
      rootModel

    useEffect(() => {
      if (firstLoad && session) {
        setFirstLoad(false)
      }
    }, [firstLoad, session])

    if (error) {
      throw error
    }

    const theme = getConf(jbrowse, 'theme')

    const colorTheme = React.useMemo(
      () => ({
        palette: {
          type: mode,
        },
      }),
      [mode],
    )

    return (
      // @ts-ignore
      <ThemeProvider theme={createJBrowseTheme(deepmerge(theme, colorTheme))}>
        <CssBaseline />
        {session ? (
          <>
            <App session={session} />
            <Suspense fallback={<div />}>
              {isAssemblyEditing ? (
                <AssemblyManager
                  rootModel={rootModel}
                  onClose={() => {
                    setAssemblyEditing(false)
                  }}
                />
              ) : null}
            </Suspense>
          </>
        ) : (
          <div>No session</div>
        )}
      </ThemeProvider>
    )
  },
)

export default JBrowse
