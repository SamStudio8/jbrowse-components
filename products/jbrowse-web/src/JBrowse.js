import { App, theme } from '@gmod/jbrowse-core/ui'
import CssBaseline from '@material-ui/core/CssBaseline'
import { ThemeProvider } from '@material-ui/core/styles'
import { observer } from 'mobx-react'
import { onSnapshot } from 'mobx-state-tree'
import { StringParam, useQueryParam } from 'use-query-params'
import React, { useEffect } from 'react'

// adapted from https://github.com/jashkenas/underscore/blob/5d8ab5e37c9724f6f1181c5f95d0020815e4cb77/underscore.js#L894-L925
function debounce(func, wait) {
  let timeout
  let result
  const later = (...args) => {
    timeout = null
    result = func(...args)
  }
  const debounced = (...args) => {
    if (timeout) {
      clearTimeout(timeout)
    }
    timeout = setTimeout(() => {
      return later(...args)
    }, wait)
    return result
  }
  debounced.cancel = () => {
    clearTimeout(timeout)
    timeout = null
  }
  return debounced
}

const JBrowse = observer(({ pluginManager }) => {
  const [sessionId] = useQueryParam('session', StringParam)

  const { rootModel } = pluginManager
  const { session, error } = rootModel || {}

  useEffect(() => {
    function updateLocalSession(snapshot) {
      // TODOSESSION have warning when reaching full local storage, have them manually delete instead of auto delete
      // sessionstorage over local storage, make save button that just transfers from session to local
      // autosave one in localstorage called local-autosaved that continuously updates on the most recent change
      console.log('updating-local-storage', sessionId)
      if (sessionId?.startsWith('local-'))
        localStorage.setItem(sessionId, JSON.stringify(snapshot))
    }

    let disposer = () => {}
    if (session) {
      const updater = debounce(updateLocalSession, 400)
      const snapshotDisposer = onSnapshot(session, updater)
      disposer = () => {
        snapshotDisposer()
        updater.cancel()
      }
    }
    return disposer
  }, [session, sessionId])

  if (error) {
    throw new Error(error)
  }

  return <App session={rootModel.session} />
})

export default props => {
  return (
    <ThemeProvider theme={theme}>
      <CssBaseline />
      <JBrowse {...props} />
    </ThemeProvider>
  )
}
