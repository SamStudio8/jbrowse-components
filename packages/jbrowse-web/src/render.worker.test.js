import JBrowse from './JBrowse'
import { renderRegion, freeSessionResources } from './render.worker'

const jbrowse = new JBrowse().configure()

test('can render a single region with Pileup + BamAdapter', async () => {
  const testprops = {
    region: { assembly: 'volvox', refName: 'ctgA', start: 0, end: 800 },
    sessionId: 'knickers the cow',
    adapterType: 'BamAdapter',
    adapterConfig: {
      _configId: '7Hc9NkuD4x',
      type: 'BamAdapter',
      bamLocation: {
        path: require.resolve('../public/test_data/volvox-sorted.bam'),
      },
      index: {
        _configId: 'sGW8va26pr',
        location: {
          path: require.resolve('../public/test_data/volvox-sorted.bam.bai'),
        },
      },
    },
    rendererType: 'PileupRenderer',
    renderProps: {},
  }

  const result = await renderRegion(jbrowse.pluginManager, testprops)
  expect(Object.keys(result)).toEqual(['featureJSON', 'html'])
  expect(result.featureJSON.length).toBe(93)
  expect(result.html).toMatchSnapshot()

  expect(freeSessionResources('knickers the cow')).toBe(1)
})
