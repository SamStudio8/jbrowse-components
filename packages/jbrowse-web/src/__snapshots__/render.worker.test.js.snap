// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can render a single region with Pileup + BamAdapter 1`] = `"<div class=\\"PileupRendering\\" data-reactroot=\\"\\"><div class=\\"feature\\" data-feature-id=\\"2135426977\\">2135426977</div><div class=\\"feature\\" data-feature-id=\\"-487480200\\">-487480200</div><div class=\\"feature\\" data-feature-id=\\"1390838279\\">1390838279</div><div class=\\"feature\\" data-feature-id=\\"1150161791\\">1150161791</div><div class=\\"feature\\" data-feature-id=\\"-432814408\\">-432814408</div><div class=\\"feature\\" data-feature-id=\\"-1327260758\\">-1327260758</div><div class=\\"feature\\" data-feature-id=\\"904949709\\">904949709</div><div class=\\"feature\\" data-feature-id=\\"-1573216292\\">-1573216292</div><div class=\\"feature\\" data-feature-id=\\"1542598533\\">1542598533</div><div class=\\"feature\\" data-feature-id=\\"208318216\\">208318216</div><div class=\\"feature\\" data-feature-id=\\"-526084999\\">-526084999</div><div class=\\"feature\\" data-feature-id=\\"-1402357424\\">-1402357424</div><div class=\\"feature\\" data-feature-id=\\"230630259\\">230630259</div><div class=\\"feature\\" data-feature-id=\\"-1240580270\\">-1240580270</div><div class=\\"feature\\" data-feature-id=\\"-126415083\\">-126415083</div><div class=\\"feature\\" data-feature-id=\\"-924959209\\">-924959209</div><div class=\\"feature\\" data-feature-id=\\"1213790853\\">1213790853</div><div class=\\"feature\\" data-feature-id=\\"1480692980\\">1480692980</div><div class=\\"feature\\" data-feature-id=\\"1806471452\\">1806471452</div><div class=\\"feature\\" data-feature-id=\\"1114949599\\">1114949599</div><div class=\\"feature\\" data-feature-id=\\"-624523461\\">-624523461</div><div class=\\"feature\\" data-feature-id=\\"941345053\\">941345053</div><div class=\\"feature\\" data-feature-id=\\"543331399\\">543331399</div><div class=\\"feature\\" data-feature-id=\\"-970365713\\">-970365713</div><div class=\\"feature\\" data-feature-id=\\"1478804620\\">1478804620</div><div class=\\"feature\\" data-feature-id=\\"710214930\\">710214930</div><div class=\\"feature\\" data-feature-id=\\"-1590347580\\">-1590347580</div><div class=\\"feature\\" data-feature-id=\\"1615231956\\">1615231956</div><div class=\\"feature\\" data-feature-id=\\"-999079398\\">-999079398</div><div class=\\"feature\\" data-feature-id=\\"212318835\\">212318835</div><div class=\\"feature\\" data-feature-id=\\"21405281\\">21405281</div><div class=\\"feature\\" data-feature-id=\\"298547082\\">298547082</div><div class=\\"feature\\" data-feature-id=\\"-632885881\\">-632885881</div><div class=\\"feature\\" data-feature-id=\\"230559949\\">230559949</div><div class=\\"feature\\" data-feature-id=\\"1358701805\\">1358701805</div><div class=\\"feature\\" data-feature-id=\\"-1150282940\\">-1150282940</div><div class=\\"feature\\" data-feature-id=\\"-584714253\\">-584714253</div><div class=\\"feature\\" data-feature-id=\\"-1543509453\\">-1543509453</div><div class=\\"feature\\" data-feature-id=\\"49754882\\">49754882</div><div class=\\"feature\\" data-feature-id=\\"1697419078\\">1697419078</div><div class=\\"feature\\" data-feature-id=\\"-467929454\\">-467929454</div><div class=\\"feature\\" data-feature-id=\\"-1903522961\\">-1903522961</div><div class=\\"feature\\" data-feature-id=\\"1449206326\\">1449206326</div><div class=\\"feature\\" data-feature-id=\\"748240882\\">748240882</div><div class=\\"feature\\" data-feature-id=\\"93246257\\">93246257</div><div class=\\"feature\\" data-feature-id=\\"911993950\\">911993950</div><div class=\\"feature\\" data-feature-id=\\"93942220\\">93942220</div><div class=\\"feature\\" data-feature-id=\\"-54442761\\">-54442761</div><div class=\\"feature\\" data-feature-id=\\"-1682306516\\">-1682306516</div><div class=\\"feature\\" data-feature-id=\\"881925482\\">881925482</div><div class=\\"feature\\" data-feature-id=\\"531931475\\">531931475</div><div class=\\"feature\\" data-feature-id=\\"1004747877\\">1004747877</div><div class=\\"feature\\" data-feature-id=\\"-724923975\\">-724923975</div><div class=\\"feature\\" data-feature-id=\\"640474988\\">640474988</div><div class=\\"feature\\" data-feature-id=\\"-783205836\\">-783205836</div><div class=\\"feature\\" data-feature-id=\\"2112054817\\">2112054817</div><div class=\\"feature\\" data-feature-id=\\"431313551\\">431313551</div><div class=\\"feature\\" data-feature-id=\\"-1280931855\\">-1280931855</div><div class=\\"feature\\" data-feature-id=\\"679424769\\">679424769</div><div class=\\"feature\\" data-feature-id=\\"-1294921403\\">-1294921403</div><div class=\\"feature\\" data-feature-id=\\"-397495927\\">-397495927</div><div class=\\"feature\\" data-feature-id=\\"1514689949\\">1514689949</div><div class=\\"feature\\" data-feature-id=\\"-1676418289\\">-1676418289</div><div class=\\"feature\\" data-feature-id=\\"1846631799\\">1846631799</div><div class=\\"feature\\" data-feature-id=\\"1272673179\\">1272673179</div><div class=\\"feature\\" data-feature-id=\\"1882240796\\">1882240796</div><div class=\\"feature\\" data-feature-id=\\"-233455955\\">-233455955</div><div class=\\"feature\\" data-feature-id=\\"529426168\\">529426168</div><div class=\\"feature\\" data-feature-id=\\"68768545\\">68768545</div><div class=\\"feature\\" data-feature-id=\\"-1578361879\\">-1578361879</div><div class=\\"feature\\" data-feature-id=\\"-1730672711\\">-1730672711</div><div class=\\"feature\\" data-feature-id=\\"687062757\\">687062757</div><div class=\\"feature\\" data-feature-id=\\"-1055583960\\">-1055583960</div><div class=\\"feature\\" data-feature-id=\\"374397043\\">374397043</div><div class=\\"feature\\" data-feature-id=\\"1401207929\\">1401207929</div><div class=\\"feature\\" data-feature-id=\\"-1861902917\\">-1861902917</div><div class=\\"feature\\" data-feature-id=\\"-541132950\\">-541132950</div><div class=\\"feature\\" data-feature-id=\\"1559390564\\">1559390564</div><div class=\\"feature\\" data-feature-id=\\"1086146284\\">1086146284</div><div class=\\"feature\\" data-feature-id=\\"-1706997693\\">-1706997693</div><div class=\\"feature\\" data-feature-id=\\"-1199599379\\">-1199599379</div><div class=\\"feature\\" data-feature-id=\\"53128272\\">53128272</div><div class=\\"feature\\" data-feature-id=\\"1629308221\\">1629308221</div><div class=\\"feature\\" data-feature-id=\\"-517500600\\">-517500600</div><div class=\\"feature\\" data-feature-id=\\"1410870858\\">1410870858</div><div class=\\"feature\\" data-feature-id=\\"-866123030\\">-866123030</div><div class=\\"feature\\" data-feature-id=\\"1046677334\\">1046677334</div><div class=\\"feature\\" data-feature-id=\\"750545343\\">750545343</div><div class=\\"feature\\" data-feature-id=\\"-1157339540\\">-1157339540</div><div class=\\"feature\\" data-feature-id=\\"1079933533\\">1079933533</div><div class=\\"feature\\" data-feature-id=\\"-262734143\\">-262734143</div><div class=\\"feature\\" data-feature-id=\\"-94743417\\">-94743417</div><div class=\\"feature\\" data-feature-id=\\"-1223356361\\">-1223356361</div></div>"`;
